rbd.c

```diff
@@ -768,6 +875,11 @@ static int tcmu_rbd_unlock(struct tcmu_device *dev)
 static int tcmu_rbd_lock(struct tcmu_device *dev, uint16_t tag)
 {
 	struct tcmu_rbd_state *state = tcmur_dev_get_private(dev);
+#if !defined LIBRBD_SUPPORTS_GETADDRS && defined RBD_LOCK_ACQUIRE_SUPPORT
```

Comment:
Nit: I wouldn't worry about adding this alternative approach. We will work to backport the new API to Octopus and Nautilus so we just need the guards on `LIBRADOS_SUPPORTS_GETADDRS` (not `LIBRBD_SUPPORTS_GETADDRS`) to prevent compile errors.

Reply:
Yeah, okay, let's keep it just for the case for some one who are using the old ceph versions without the backporting.

Reply:
The `LIBRADOS_SUPPORTS_GETADDRS` support will be backported through nautilus so for the sake of code-clarity and simplicity, we shouldn't have to worry about alternative approaches (especially since the alternative has a potential race condition).

Reply:
Yeah, that's cool.

For the alternative approach, there should have the potential race condition only in theory when one tcmu-runner node gettting the same nonce twice.

This alternative approach here is a little different from the one we disscussed in the mail before, it will do:

1, after the local tcmu-runner node successfully getting the exclusive lock, it will try to get the lock owner and save it as `lock_owner1`.

2, then check if the local tcmu-runner node is the lock owner. If not, it means local tcmu-runner node had lost the lock, so we cannot trust the `lock_owner1`, then just skip saving the entity addr to local cache arrary and no need to do the following steps.

3, If local node is still the lock owner in Step2, then try to get the lock owner again and save it as `lock_owner2`.

4, only when `lock_owner1` == `lock_owner2`, that means we never lost the lock during Step1. Then we can trust the `lock_owner1` and save it to local cache arrary.

In Step4, if the IP address is foreign tcmu-runner node's, that means during both Step1 and Step3, the lock is acquired by the same foreign tcmu-runner node. and in Step2 the local node acquired it again.  And at the same time in Step1 and Step3 the `nonces` are the same, IMO this should happen only in theory.

Let's just assume the above case won't happen, that means during Step1, Step2 and Step3, the local node never loses the lock.

Reply:
> In Step4 if the IP address is local tcmu-runner nodes's, and it has lost the lock once, then in local node it also got the same nonce twice.

For what it's worth, the nonce will only change if you disconnect from the cluster and reconnect -- which hopefully isn't happening while you are within this method.

Reply:
Yeah, it is. And the Step2 is a barrier.

Since it will reopen/reconnect the image if the local/foreign node will re-acquire the lock. 

If in Step2 it says the local node is the lock owner, it means local node never lost the lock or has re-acquired the lock since Step1, if it is the later case, the reconnection must already be done.

No matter the IP is local or foreign node's in `lock_owner1` and `lock_owner2`. Without considering the case that one node will get a same `nonce` value with previous connection,  if Step4 says `lock_owner1` and `lock_owner2` are the same, we can say that the local node never lost the lock.

I have run the `nonce` generator code for round 1 thousand times, didn't get a same nonce with the existings. WIth this approach we may miss removing some blacklist entries, but it shouldn't matter.